### PR TITLE
Fixes to Array class

### PR DESCRIPTION
**Pull request by adbridge**
Add default initialisers in the Array constructor for internal parameters
This fixes github issue 67

Added a test to the push_back() method to check that the init() method has
previously been invoked before allowing elements to be added.
This fixes github issue 71

### CODE REVIEW

**Comment by adbridge** [comment]
@bogdanm  @0xc0170 Code changes

File: core-util/Array.h [hunk 1]
```diff
@@ -131,10 +138,18 @@ class Array {
...
       * @param new_element element to add
-      * @returns true if the element was added, false otherwise (out of memory)
+      * @returns true if the element was added, false otherwise (out of memory/uninitialised)
       */
     bool push_back(const T& new_element) {
         unsigned idx = _elements, prev_capacity = _capacity;
+
+        // element_size is calculated in the init method, thus this can be tested to determine
+        // whether or not init has been called previously.
+        if (_element_size == 0 ){
```
Comment (0xc0170) [inline core-util/Array.h]: style :-) `` if (_element_size == 0) { ``

File: core-util/Array.h [hunk 2]
```diff
@@ -45,6 +46,12 @@ class Array {
     /** Create a new array
       */
     Array(): _head(NULL) {
+        _grow_capacity = 0;
+        _alloc_traits = {0};
+        _alignment = MBED_UTIL_POOL_ALLOC_DEFAULT_ALIGN;
+        _capacity = 0;
+        _elements = 0;
+        _element_size = 0;
```
Comment (0xc0170) [inline core-util/Array.h]: In this case it is a matter of style, initializer list or might be better to use c++11 initializers :question:

**Comment by 0xc0170** [comment]
@adbridge Please edit the commit message, it was not updated to the latest change to this change-set

**Comment by adbridge** [comment]
@0xc0170 I don't understand your last comment: The commit message says this:

Add default initialisers in the Array constructor for internal parame… …
adbridge committed 5 days ago

…ters.

This fixes github issue 67

Added a test to the push_back() method to check that the init() method has
previously been invoked before allowing elements to be added.
This fixes github issue 71

Which is the correct commit message ?

**Comment by 0xc0170** [comment]
> @0xc0170 I don't understand your last comment: The commit message says this:

the commit message - Added a test - I understood this one refers to previous test implementation which was part of this PR. I see now that it refers to additional check in the implementation, which I misunderstood. A tip: You can use #67 for instance to reference the issue (creates a link).

File: core-util/Array.h (hunk 2 above)
Comment (adbridge) [inline core-util/Array.h]: It didn't like me putting the defaults into the initialiser list , get -Wreorder warnings due to values being set in the init function. As this is all likely to change soon anyway as there are a number of optimisations we would like to do on this code, may as well leave it as is for now.

File: core-util/Array.h (hunk 1 above)
Comment (adbridge) [inline core-util/Array.h]: fixed

**Comment by 0xc0170** [comment]
@bogdanm Pls review

**Comment by bogdanm** [comment]
LGTM, +1. This is the least we can do for now.

**Pull request merged**